### PR TITLE
DS-4946 add negative user sampling to modeling in collie

### PR DESCRIPTION
**Pull request by emmafgrimes**
## Describe the big picture - why are we doing this?
This PR adds the ability to do negative user sampling in the `BasePipeline` and the loss files. Again, I did not edit the `changelog` nor `_version`, that will come in the next PR when all the parts of negative user sampling have been added!

## How is this tested?
On an EC2 instance in Docker, all tests pass.

## Pull Request Checklist
 - [x] Pull request includes a description of why we are doing this
 - [ ] ``CHANGELOG`` has been updated
 - [ ] Version in ``_version.py`` has been updated
 - [x] All tests in the ``tests`` folder pass with a local build
 - [ ] ``README`` has been updated (if applicable)
 - [ ] Documentation in ``docs`` has been updated (if applicable)
 - [ ] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)
 - [x] Docker image can be built using ``docker build -t collie .``

### CODE REVIEW

File: collie/loss/metadata_utils.py [hunk 1]
```diff
@@ -4,19 +4,20 @@
...
-    positive_items: torch.tensor,
-    negative_items: torch.tensor,
+    positive_ids: torch.tensor,
+    negative_ids: torch.tensor,
     metadata: Optional[Dict[str, torch.tensor]],
     metadata_weights: Optional[Dict[str, float]],
 ) -> torch.tensor:
     """
-    Helper function to calculate the ideal score difference between the positive and negative items.
+    Helper function to calculate the ideal score difference between the positive and negative ids.
```
Comment (nathancooperjones) [inline collie/loss/metadata_utils.py]: Same nitpicky `ids -> IDs` comment here :)

File: collie/model/base/base_pipeline.py [hunk 2]
```diff
@@ -179,14 +179,27 @@ def __init__(self,
...
+                if (
+                    hasattr(self.train_loader, 'negative_sample_type')
+                    or hasattr(self.val_loader, 'negative_sample_type')
+                ):
+                    assert self.train_loader.negative_sample_type == \
+                        self.val_loader.negative_sample_type, (
+                            'Training and val ``negative_sample_type`` property must both equal '
+                            '``item`` or both equal ``user``, not: '
+                            f'{self.train_loader.negative_sample_type}'
+                            f' and {self.val_loader.negative_sample_type}, respectively.'
```
Comment (nathancooperjones) [inline collie/model/base/base_pipeline.py]: I think it's okay to just say something along the lines of: 

```suggestion
                            'Training and val ``negative_sample_type`` property must both equal: '
                            f'{self.train_loader.negative_sample_type} != '
                            f'{self.val_loader.negative_sample_type}.'
```

File: collie/model/base/base_pipeline.py [hunk 3]
```diff
@@ -179,14 +179,27 @@ def __init__(self,
...
                     hasattr(self.train_loader, 'num_negative_samples')
                     or hasattr(self.val_loader, 'num_negative_samples')
                 ):
                     num_negative_samples_error = (
                         'Training and val ``num_negative_samples`` property must both equal ``1``'
-                        f' or both be greater than ``1``, not: {self.train_loader.num_items} and'
-                        f' {self.val_loader.num_items}, respectively.'
+                        ' or both be greater than ``1``, not: '
+                        f'{self.train_loader.num_negative_samples}'
+                        f' and {self.val_loader.num_negative_samples}, respectively.'
```
Comment (nathancooperjones) [inline collie/model/base/base_pipeline.py:201]: Good catch!

File: tests/test_model.py [hunk 4]
```diff
@@ -95,6 +96,9 @@ def test_mismatched_train_and_val_loaders(train_val_implicit_data, change_to_mak
     elif change_to_make == 'num_items':
         train.num_items = 1
         val.num_items = 2
+    elif change_to_make == 'negative_sample_type':
+        train.negative_sample_type = 'item'
+        val.negative_sample_type = 'user'
```
Comment (nathancooperjones) [inline tests/test_model.py:101]: Can we also add that test we talked about offline which is just the copy of the `test_implicit_model` test in `model.py`. This will unfortunately require creating a billion more movielens + models fixtures (sorry) 😬

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
This looks great - yabba-dabba-doo!!! 🐕‍🦺 

Also the small one-line bug in `cross_validation.py`.

File: collie/loss/metadata_utils.py (hunk 1 above)
Comment (emmafgrimes) [reply collie/loss/metadata_utils.py]: Ugh! And I even updated it elsewhere in the code 🤦‍♀️  thanks for catching.

**Comment by emmafgrimes** [comment]
Thanks @nathancooperjones ! Back to you 🚀

File: tests/fixtures/movielens_fixtures.py [hunk 5]
```diff
@@ -50,6 +50,19 @@ def movielens_implicit_interactions(movielens_implicit_df):
...
 
 
+@pytest.fixture(scope='session')
+def movielens_implicit_interactions_user_negative_sample_type(
+    movielens_implicit_df
+):
+    return Interactions(users=movielens_implicit_df['user_id'],
+                        items=movielens_implicit_df['item_id'],
+                        ratings=movielens_implicit_df['rating'],
+                        negative_sample_type='user',
```
Comment (nathancooperjones) [inline tests/fixtures/movielens_fixtures.py:60]: [Optional] Maybe in `movielens_implicit_interactions` we can be super explicit and set `negative_sample_type='item'`?

**Review by nathancooperjones — APPROVED** [review]
🎉

File: tests/fixtures/movielens_fixtures.py (hunk 5 above)
Comment (emmafgrimes) [reply tests/fixtures/movielens_fixtures.py:60]: will do in the next round :)
